### ISSUE TITLE
`torch.where` on MPS fails to broadcast identically to other backends / numpy

### ISSUE THREAD

**Issue by aslvrstn** [pytorch/pytorch#86239]
### 🐛 Describe the bug

```
import torch as t

t.where(t.ones(10, 10).to("mps").bool(), t.ones(2, 10, 10).to("mps"), t.ones(10, 10).to("mps"))
```

this succeeds on CPU and fails with

> RuntimeError: 0'th index 2 of x tensor does not match the other tensors

on MPS

I believe I have a fix which I will link immediately after creating this issue.

### Versions

(composition) alex@Alexs-MacBook-Pro-2 mlscratchpad % python ../pytorch/torch/utils/collect_env.py
Collecting environment information...
PyTorch version: 1.13.0.dev20221004
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.201)
CMake version: version 3.24.2
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.13.0.dev20221004
[pip3] torchaudio==0.13.0.dev20221004
[pip3] torchvision==0.15.0.dev20221004
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] numpy-base                1.23.1          py310h742c864_0
[conda] torch                     1.13.0.dev20221004          pypi_0    pypi
[conda] torchaudio                0.13.0.dev20221004          pypi_0    pypi
[conda] torchvision               0.15.0.dev20221004          pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by aslvrstn** [comment]
I just noticed that this fixes another worse bug in the process. On some shapes `t.where` will successfully return different shaped arrays rather than just erroring.

```
>>> t.where(t.tensor(False), t.zeros(2), t.zeros((1, 1)))
tensor([[0., 0.]])
```

```
>>> t.where(t.tensor(False).to("mps"), t.zeros(2).to("mps"), t.zeros((1, 1)).to("mps"))
tensor([[0.],
        [0.]], device='mps:0')
```


Found this with a hypothesis test:
```
from hypothesis.extra.numpy import mutually_broadcastable_shapes
import numpy as np
import hypothesis
import hypothesis.strategies
from numpy.testing import assert_array_equal

import torch as t


@hypothesis.strategies.composite
def where_inputs(draw, xy_dtype = int):
    shapes, result_shape = draw(mutually_broadcastable_shapes(num_shapes=3))
    cond = draw(hypothesis.extra.numpy.arrays(bool, shape=shapes[0]))
    x = draw(hypothesis.extra.numpy.arrays(xy_dtype, shape=shapes[1]))
    y = draw(hypothesis.extra.numpy.arrays(xy_dtype, shape=shapes[2]))
    return cond, x, y


class TestMPS(unittest.TestCase):
    @given(where_inputs())
    def test_where_consistent(self, where_inputs):
        dev = "mps"

        cond, x, y = where_inputs
        t_cond, t_x, t_y = t.from_numpy(cond), t.from_numpy(x), t.from_numpy(y)

        np_out = np.where(cond, x, y)
        torch_cpu_out = t.where(t_cond, t_x, t_y)
        torch_mps_out = t.where(t_cond.to(dev), t_x.to(dev), t_y.to(dev))

        assert_array_equal(np_out, torch_cpu_out.numpy())
        assert_array_equal(np_out, torch_mps_out.cpu().numpy())
```

which fails at
```
>       assert_array_equal(np_out, torch_mps_out.cpu().numpy())
E       AssertionError:
E       Arrays are not equal
E
E       (shapes (1, 2), (2, 1) mismatch)
E        x: array([[0, 0]])
E        y: array([[0],
E              [0]])
E       Falsifying example: test_where_consistent(
E           where_inputs=(array(False), array([0, 0]), array([[0]])),
E           self=<fuzz_mps.TestMPS testMethod=test_where_consistent>,
E       )

fuzz_mps.py:38: AssertionError
```


I'm not familiar with torch's unit test setup -- do you have a place that a test like this would live nicely?

**Comment by aslvrstn** [comment]
@kulinseth I got curious about the setup for making this test run against all backends, and I ended up with https://github.com/pytorch/pytorch/pull/86977 . The `where` test in this PR successfully detects this bug (when I revert the fix) but also tests across other backends, I think.

I'm struggling to get CI to pick up that it should run a mac CI run for this PR though, which makes me think I'm missing something. Does this seem like a reasonable direction?
